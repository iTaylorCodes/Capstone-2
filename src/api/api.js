import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";
const BASE_PLACES_URL = process.env.REACT_APP_PLACES_URL;
const BASE_PHOTO_URL = process.env.REACT_APP_PHOTO_URL;
const PROXY_URL = process.env.REACT_APP_PROXY_URL;
const BASE_GEOCODE_URL = process.env.REACT_APP_GEOCODE_URL;
const BASE_WALKSCORE_URL = process.env.REACT_APP_WALKSCORE_URL;
const BASE_HOTELS_URL = process.env.REACT_APP_BASE_HOTELS_URL;

/** API Class.
 *
 * Static class tying together methods used to
 *    get/send to the backend database API.
 */

class NeighborhoodApi {
  // The token for interacting with the API will be stored here.
  static token;

  static async request(endpoint, data = {}, method = "get") {
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${NeighborhoodApi.token}` };
    const params = method === "get" ? data : {};

    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.response);
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    }
  }

  // Individual API routes

  /** Get the current user by username */
  static async getCurrentUser(username) {
    let res = await this.request(`users/${username}`);
    return res.user;
  }

  /** Get token for login from username, password */
  static async login(data) {
    let res = await this.request(`auth/token`, data, "post");
    return res.token;
  }

  /** Signup */
  static async signup(data) {
    let res = await this.request(`auth/register`, data, "post");
    return res.token;
  }

  /** Update user profile */
  static async updateProfile(username, data) {
    let res = await this.request(`users/${username}`, data, "patch");
    return res.user;
  }

  /** Delete user profile */
  static async deleteProfile(username) {
    await this.request(`users/${username}`, {}, "delete");
  }

  /** Add property to user favorites */
  static async favoriteProperty(username, propertyZpid) {
    await this.request(`users/${username}/${propertyZpid}`, {}, "post");
  }

  /** Remove property from user favorites */
  static async unFavoriteProperty(username, propertyZpid) {
    await this.request(`users/${username}/${propertyZpid}`, {}, "delete");
  }
}

/** API Class.
 *
 * Static class to get city images from Google places API.
 */

class PlacesApi {
  static async getImage(cityState) {
    const placesUrl = `${BASE_PLACES_URL}/json?input=${cityState}&key=${process.env.REACT_APP_PLACES_API_KEY}&inputtype=textquery&fields=name,photos`;

    const initialPlacesRequest = await axios
      .get(PROXY_URL + placesUrl)
      .catch(console.error);

    const photoRef =
      initialPlacesRequest?.data?.candidates?.[0]?.photos?.[0]?.photo_reference;

    if (!photoRef) throw Error("Image not found.");

    const imageLookupUrl = `${BASE_PHOTO_URL}?photoreference=${photoRef}&key=${process.env.REACT_APP_PLACES_API_KEY}&maxwidth=700&maxheight=700`;

    const imageUrlQuery = await fetch(PROXY_URL + imageLookupUrl)
      .then((r) => r.blob())
      .catch(console.error);

    const image = URL.createObjectURL(imageUrlQuery);

    return image;
  }
}

/** API Class.
 *
 * Static class get walk scores from Walk Score API.
 */

class WalkScoreApi {
  static async getScores(city) {
    const geocodeUrl = `${BASE_GEOCODE_URL}?key=${process.env.REACT_APP_GEOCODE_API_KEY}&location=${city}`;

    const geocodeRequest = await axios.get(geocodeUrl);

    const lat = geocodeRequest.data.results[0].locations[0].latLng.lat;
    const long = geocodeRequest.data.results[0].locations[0].latLng.lng;

    if (!lat) throw Error("City not found.");
    if (!long) throw Error("City not found.");

    const requestUrl = `${BASE_WALKSCORE_URL}/score?format=json&address=${city}&lat=${lat}&lon=${long}&transit=1&bike=1&wsapikey=${process.env.REACT_APP_WALKSCORE_API_KEY}`;

    const scoreRequest = (await axios.get(PROXY_URL + requestUrl)).data;

    const walkScore = scoreRequest.walkscore
      ? {
          score: scoreRequest.walkscore,
          description: scoreRequest.description,
        }
      : null;

    const transitScore = scoreRequest.transit
      ? {
          score: scoreRequest.transit.score,
          description: scoreRequest.transit.description,
        }
      : null;

    const bikeScore = scoreRequest.bike
      ? {
          score: scoreRequest.bike.score,
          description: scoreRequest.bike.description,
        }
      : null;

    const scores = {
      walkScore: walkScore,
      transitScore: transitScore,
      bikeScore: bikeScore,
    };

    return scores;
  }
}

/** API Class.
 *
 * Static class to get hotel data from Api Dojo Hotels API.
 */

class HotelsApi {
  static headers = {
    "x-rapidapi-host": "hotels4.p.rapidapi.com",
    "x-rapidapi-key": process.env.REACT_APP_X_RAPIDAPI_KEY,
  };

  /** Get cityId for hotel search */
  static async getCityId(city) {
    const requestUrl = `${BASE_HOTELS_URL}/locations/v2/search?query=${city}`;

    const cityRequest = await axios
      .get(PROXY_URL + requestUrl, { headers: HotelsApi.headers })
      .catch(console.error);

    const cityId =
      cityRequest?.data?.suggestions?.[0]?.entities?.[0]?.destinationId;

    if (!cityId) throw Error("City not found.");

    return cityId;
  }

  /** Get list of hotels */
  static async getHotels(city) {
    const cityId = await HotelsApi.getCityId(city);

    const requestUrl = `${BASE_HOTELS_URL}/properties/list?destinationId=${cityId}&pageNumber=1&pageSize=10&checkIn=2021-01-01&checkOut=2021-01-10&adults1=1`;

    const hotelsRequest = await axios
      .get(PROXY_URL + requestUrl, { headers: HotelsApi.headers })
      .catch(console.error);

    const hotels = hotelsRequest?.data?.data?.body?.searchResults?.results;

    return hotels;
  }

  /** Get hotel details */
  static async getHotel(hotelId) {
    const requestUrl = `${BASE_HOTELS_URL}/properties/get-details?id=${hotelId}`;

    const hotelRequest = await axios
      .get(PROXY_URL + requestUrl, { headers: HotelsApi.headers })
      .catch(console.error);

    const hotel = hotelRequest?.data?.data?.body;

    return hotel;
  }
}

export { NeighborhoodApi, PlacesApi, WalkScoreApi, HotelsApi };
