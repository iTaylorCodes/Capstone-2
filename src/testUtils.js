import UserContext from "./auth/UserContext";

const testUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
};

const UserProvider = ({ children, currentUser = testUser }) => (
  <UserContext.Provider value={{ currentUser }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };
