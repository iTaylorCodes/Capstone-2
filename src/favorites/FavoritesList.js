import { useState, useContext, useEffect } from "react";
import { HotelsApi } from "../api/api";
import UserContext from "../auth/UserContext";
import FavoriteHotelCard from "./FavoriteHotelCard";

/** Favorites List.
 *
 * Component dedicated to showing a list of the users favorited Hotels.
 *
 * Renders multiple HotelCard components,
 * passing in data from HotelsApi based on hotelIds saved by the user.
 *
 * Rendered at /favorites.
 */
const FavoritesList = () => {
  const { currentUser } = useContext(UserContext);
  const [favHotels, setFavHotels] = useState([]);
  const userFavs = currentUser.favoritedProperties;

  useEffect(() => {
    let isSubscribed = true;
    const fetchHotel = async () => {
      try {
        if (isSubscribed) {
          for (let val of userFavs) {
            const hotel = await HotelsApi.getHotel(val);
            setFavHotels([...favHotels, hotel]);
          }
        }
      } catch (e) {
        console.error(e);
      }
    };
    fetchHotel();
    return () => (isSubscribed = false);
  }, [userFavs]);

  return (
    <div className="FavoritesList">
      <div className="container col-12 col-sm-9 col-md-8 col-lg-7">
        <div className="row mt-3 ms-1 me-1">
          <div className="col-12">
            <h1>Favorited Hotels</h1>
          </div>
        </div>
        {favHotels.length ? (
          <div className="row g-0 d-flex justify-content-center container">
            {favHotels.map((h) => {
              return (
                <div
                  className="col-10 col-sm-12 col-md-6 col-xxl-4"
                  key={h.pdpHeader.hotelId}
                >
                  <FavoriteHotelCard hotel={h} />
                </div>
              );
            })}
          </div>
        ) : (
          <div>
            <h3 className="text-danger d-flex justify-content-center text-center">
              No Favorites Saved Yet.
            </h3>
            <p className="text-danger d-flex justify-content-center text-center">
              Try adding to your favorites after completing a neighborhood
              search.
            </p>
          </div>
        )}
      </div>
    </div>
  );
};

export default FavoritesList;
