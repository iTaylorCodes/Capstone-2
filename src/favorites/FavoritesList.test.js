import FavoritesList from "./FavoritesList";
import { render } from "@testing-library/react";
import { UserProvider } from "../testUtils";
import { HotelsApi } from "../api/api";

const hotels = [
  {
    pdpHeader: { hotelId: "12345" },
    propertyDescription: {
      name: "Tester Hotel",
      address: {
        cityName: "Los Angeles",
        fullAddress: "1234 Tester Way, Los Angeles, CA 12345",
      },
    },
    starRating: "5",
  },
];

const hotelSpy = jest.spyOn(HotelsApi, "getHotel");
hotelSpy.mockReturnValue(hotels);

it("renders without crashing", () => {
  render(
    <UserProvider>
      <FavoritesList />
    </UserProvider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <UserProvider>
      <FavoritesList />
    </UserProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
