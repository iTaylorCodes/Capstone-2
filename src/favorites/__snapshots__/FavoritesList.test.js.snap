// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="FavoritesList"
  >
    <div
      class="container col-12 col-sm-9 col-md-8 col-lg-7"
    >
      <div
        class="row mt-3 ms-1 me-1"
      >
        <div
          class="col-12"
        >
          <h1>
            Favorited Hotels
          </h1>
        </div>
      </div>
      <div>
        <h3
          class="text-danger d-flex justify-content-center text-center"
        >
          No Favorites Saved Yet.
        </h3>
        <p
          class="text-danger d-flex justify-content-center text-center"
        >
          Try adding to your favorites after completing a neighborhood search.
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
